File: SWEA/D5/3421. 수제 버거 장인/수제 버거 장인.java
import java.util.*;
import java.io.*;

public class Solution
{
	static int cnt = 0;
	static void hamburger(boolean visited[], ArrayList<Integer> food[], int idx, int N)
	{
		if(idx > 1)
		{
			for(int i = 0; i < idx; i++)
			{
				if(!visited[i]) continue;
				for(int f : food[i])
				{
					if(visited[i] && visited[f]) return;
				}
			}
		}
		
		if(idx == visited.length)
		{
			cnt++;
			return;
		}
		
		visited[idx] = true;
		hamburger(visited, food, idx + 1, N);
		visited[idx] = false;
		hamburger(visited, food, idx + 1, N);
	}
	
	public static void main(String[] args) throws Exception 
	{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	    StringBuilder sb = new StringBuilder();
		int T = Integer.parseInt(br.readLine());	
	    for(int tc = 1; tc <= T; ++tc)
	    {
	    	StringTokenizer st = new StringTokenizer(br.readLine());
	    	int N = Integer.parseInt(st.nextToken());
	    	int M = Integer.parseInt(st.nextToken());
	    	
	    	ArrayList<Integer> food[] = new ArrayList[N + 1];
	    	boolean visited[] = new boolean[N + 1];
	    	for(int i = 0; i < food.length; i++)
	    	{
	    		food[i] = new ArrayList<>();
	    	}
	    	for(int i = 0; i < M; i++)
	    	{
	    		st = new StringTokenizer(br.readLine());
	    		int a = Integer.parseInt(st.nextToken());
	    		int b = Integer.parseInt(st.nextToken());
	    		food[a].add(b);
	    		food[b].add(a);
	    	}
	    	cnt = 0;
	    	hamburger(visited, food, 1, N);
	    	sb.append("#").append(tc).append(" ").append(cnt).append("\n");
	    }
	    System.out.println(sb);
	}
}

File: SWEA/D5/3421. 수제 버거 장인/수제 버거 장인.cpp
#include <iostream>
#include <vector>
#include <string.h>
#include <string>
#include <queue>
#include <algorithm>

using namespace std;

int cnt = 0;

void powerSet(vector<bool>& visited, vector<int>* food, int idx, int N, int M, int select)
{
	if (idx > 1)
	{
		for (int i = 1; i < idx; i++)
		{
			if (!visited[i]) continue;
			for (auto f : food[i])
			{
				if (visited[i] && visited[f]) return;
			}
		}
	}

	if (idx == N + 1)
	{
		if (select < 2)cnt++;
		else cnt++;
		return;
	}
	
	visited[idx] = true;
	powerSet(visited, food, idx + 1, N, M, select + 1);

	visited[idx] = false;
	powerSet(visited, food, idx + 1, N, M, select);
}

int main()
{
	int T; cin >> T;

	for (int tc = 1; tc <= T; tc++)
	{
		int N, M; cin >> N >> M;
		vector<int> food[21];
		cnt = 0;
		for (int i = 0; i < M; i++)
		{
			int a, b;
			cin >> a >> b;
			food[a].push_back(b);
		}
		vector<bool>visited(N + 1, false);
		powerSet(visited, food, 1, N, M, 0);
		cout << "#" << tc << " " << cnt << "\n";
	}
	return 0;
}